cpp/include/raft/neighbors/detail/ivf_pq_compute_similarity-inl.cuh

```diff
@@ -0,0 +1,845 @@
+  {
+    return kernel_choose_bits(pq_bits, k_max);
+  }
+
+ private:
+  static auto kernel_choose_bits(uint32_t pq_bits, uint32_t k_max)
+    -> compute_similarity_kernel_t<OutT, LutT>
+  {
+    switch (pq_bits) {
+      case 4: return kernel_try_capacity<4, kMaxCapacity>(k_max);
+      case 5: return kernel_try_capacity<5, kMaxCapacity>(k_max);
+      case 6: return kernel_try_capacity<6, kMaxCapacity>(k_max);
+      case 7: return kernel_try_capacity<7, kMaxCapacity>(k_max);
+      case 8: return kernel_try_capacity<8, kMaxCapacity>(k_max);
+      default: RAFT_FAIL("Invalid pq_bits (%u), the value must be within [4, 8]", pq_bits);
+    }
+  }
+
+  template <uint32_t PqBits, int Capacity>
+  static auto kernel_try_capacity(uint32_t k_max) -> compute_similarity_kernel_t<OutT, LutT>
+  {
+    if constexpr (Capacity > 0) {
+      if (k_max == 0 || k_max > Capacity) { return kernel_try_capacity<PqBits, 0>(k_max); }
+    }
+    if constexpr (Capacity > 1) {
+      if (k_max * 2 <= Capacity) { return kernel_try_capacity<PqBits, (Capacity / 2)>(k_max); }
+    }
+    return compute_similarity_kernel<OutT, LutT, PqBits, Capacity, PrecompBaseDiff, EnableSMemLut>;
+  }
+};
+
+// A standalone accessor function was necessary to make sure template
+// instantiation work correctly. This accessor function is not used anymore and
+// may be removed.
+template <typename OutT, typename LutT, bool PrecompBaseDiff, bool EnableSMemLut>
+auto get_compute_similarity_kernel(uint32_t pq_bits, uint32_t k_max)
+  -> compute_similarity_kernel_t<OutT, LutT>
+{
+  return compute_similarity_kernel_config<OutT, LutT, PrecompBaseDiff, EnableSMemLut>::get(pq_bits,
+                                                                                           k_max);
+}
+
+/** Estimate the occupancy for the given kernel on the given device. */
+template <typename OutT, typename LutT>
+struct occupancy_t {
+  using shmem_unit = Pow2<128>;
+
+  int blocks_per_sm = 0;
+  double occupancy  = 0.0;
+  double shmem_use  = 1.0;
+
+  inline occupancy_t() = default;
+  inline occupancy_t(size_t smem,
+                     uint32_t n_threads,
+                     compute_similarity_kernel_t<OutT, LutT> kernel,
+                     const cudaDeviceProp& dev_props)
+  {
+    RAFT_CUDA_TRY(
+      cudaOccupancyMaxActiveBlocksPerMultiprocessor(&blocks_per_sm, kernel, n_threads, smem));
+    occupancy = double(blocks_per_sm * n_threads) / double(dev_props.maxThreadsPerMultiProcessor);
+    shmem_use = double(shmem_unit::roundUp(smem) * blocks_per_sm) /
+                double(dev_props.sharedMemPerMultiprocessor);
+  }
+};
+
+template <typename OutT, typename LutT>
+struct selected {
+  compute_similarity_kernel_t<OutT, LutT> kernel;
+  dim3 grid_dim;
+  dim3 block_dim;
+  size_t smem_size;
+  size_t device_lut_size;
+};
+
+template <typename OutT, typename LutT>
+void compute_similarity_run(selected<OutT, LutT> s,
```

Comment:
Most of the code in this file  is just moved from `ivf_pq_compute_similarity.cuh`. The main difference is this section, instead of keeping on `operator()` for the `selected` struct, the kernel launcher is factored out into this function. As described by Allard, this is necessary to ensure that we can throw error message for accidental instantiation. Looks good to me.